marimbabot_hardware/src/servo_interface.cpp

```diff
@@ -0,0 +1,199 @@
+        } catch (serial::SerialException &e) {
+            ROS_ERROR_STREAM("Servo controller error: Unable to read from port " << arduino_serial.getPort());
+            ROS_ERROR_STREAM("Servo controller error: Error: " << e.what());
+            arduino_serial.close();
+            return;
+        }
+        //ROS_INFO_STREAM("Servo controller: Received response from arduino: \"" << response << "\"");
+    } while(arduino_serial.available() > 0);
+
+    // Parsing response (extract int value)
+    // remove first 2 chars and last chars (\r\n)
+    response = response.substr(2, response.length() - 4);
+    std::stringstream sstream(response);
+    int value;
+    sstream >> value;
+
+    if(sstream.fail()) {
+        ROS_WARN_STREAM("Servo controller error: Failed to parse position response from arduino");
+        return;
+    }
+
+    // Remapping pwm value to radians
+    servo_state.position = ((TOP_LIMIT - value) / ((double) TOP_LIMIT - BOTTOM_LIMIT)) * RADIAN_LIMIT;
+}
+
+void ServoInterface::write() {
+    position_joint_saturation_interface.enforceLimits(last_run_period);
+
+    // Checking if the arduino is connected
+    // Not retrying to connect here as we already tried in read()
+    if(!arduino_serial.isOpen()) {
+        return;
+    }
+
+    // Not sending commands which already where before
+    if(previous_command == servo_state.command) {
+        return;
+    }
+
+    ROS_DEBUG_STREAM("Servo controller: Sending command to arduino: " << servo_state.command);
+
+    previous_command = servo_state.command;
+
+    // Remapping pwm value from radians input
+    int command_value = TOP_LIMIT - (int) round((servo_state.command / RADIAN_LIMIT) * (TOP_LIMIT - BOTTOM_LIMIT));
+
+    // Sending command to arduino
+    std::stringstream sstream;
+    sstream << "s " << command_value << "\n";
+
+    try {
+        arduino_serial.write(sstream.str());
+        arduino_serial.flush();
+    } catch (serial::SerialException &e) {
+        ROS_ERROR_STREAM("Servo controller error: Failed to send command to arduino");
+        ROS_ERROR_STREAM("Servo controller error: Error: " << e.what());
+        arduino_serial.close();
+        return;
+    }
+
+    // Reading response from arduino
+    std::string response;
+    do {
+        try {
+            response += arduino_serial.readline();
+        } catch (serial::SerialException &e) {
+            ROS_ERROR_STREAM("Servo controller error: Unable to read from port " << arduino_serial.getPort());
+            ROS_ERROR_STREAM("Servo controller error: Error: " << e.what());
+            arduino_serial.close();
+            return;
+        }
+    } while(arduino_serial.available() > 0);
+
+    // Removing trailing \r\n
+    response = response.substr(0, response.length() - 2);
+
+    // Checking if the command was successful
+    if(response == "err_input_num") {
+        ROS_ERROR_STREAM("Servo controller error: Number unparsable or below 0");
```

Comment:
Please use `_THROTTLED`. if someone continuously sets invalid values it should not spam the console.